ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -426,208 +432,228 @@ forkBlockForging ::
-
-        -- Add the block to the chain DB
-        let noPunish = InvalidBlockPunishment.noPunishment   -- no way to punish yourself
-        -- Make sure that if an async exception is thrown while a block is
-        -- added to the chain db, we will remove txs from the mempool.
-
-        -- 'addBlockAsync' is a non-blocking action, so `mask_` would suffice,
-        -- but the finalizer is a blocking operation, hence we need to use
-        -- 'uninterruptibleMask_' to make sure that async exceptions do not
-        -- interrupt it.
-        uninterruptibleMask_ $ do
-          result <- lift $ ChainDB.addBlockAsync chainDB noPunish newBlock
-          -- Block until we have processed the block
-          mbCurTip <- lift $ atomically $ ChainDB.blockProcessed result
-
-          -- Check whether we adopted our block
-          when (mbCurTip /= SuccesfullyAddedBlock (blockPoint newBlock)) $ do
-            isInvalid <- lift $ atomically $
-              ($ blockHash newBlock) . forgetFingerprint <$>
-              ChainDB.getIsInvalidBlock chainDB
-            case isInvalid of
-              Nothing ->
-                trace $ TraceDidntAdoptBlock currentSlot newBlock
-              Just reason -> do
-                trace $ TraceForgedInvalidBlock currentSlot newBlock reason
-                -- We just produced a block that is invalid according to the
-                -- ledger in the ChainDB, while the mempool said it is valid.
-                -- There is an inconsistency between the two!
+                  (ledgerState unticked)
+
+      _ <- evaluate tickedLedgerState
+      trace $ TraceForgeTickedLedgerState currentSlot bcPrevPoint
+
+      -- Get a snapshot of the mempool that is consistent with the ledger
+      --
+      -- NOTE: It is possible that due to adoption of new blocks the
+      -- /current/ ledger will have changed. This doesn't matter: we will
+      -- produce a block that fits onto the ledger we got above; if the
+      -- ledger in the meantime changes, the block we produce here may or
+      -- may not be adopted, but it won't be invalid.
+      (mempoolHash, mempoolSlotNo) <- lift $ atomically $ do
+        snap <- getSnapshot mempool   -- only used for its tip-like information
+        let h :: ChainHash blk
+            h = castHash $ getTipHash $ snapshotState snap
+        pure (h, snapshotSlotNo snap)
+
+      let readTables = fmap castLedgerTables . roforkerReadTables forker . castLedgerTables
+
+      mempoolSnapshot <- lift $ getSnapshotFor
+                                  mempool
+                                  currentSlot
+                                  tickedLedgerState
+                                  readTables
+
+      lift $ roforkerClose forker
+
+      let txs = [ tx | (tx, _, _) <- snapshotTxs mempoolSnapshot ]
+
+      -- force the mempool's computation before the tracer event
+      _ <- evaluate (length txs)
+      _ <- evaluate mempoolHash
```

Comment:
Why switch from the snapshot's ledgerstate WHNF to the mempool hash?

Reply:
Because there is no snapshot ledgerstate anymore